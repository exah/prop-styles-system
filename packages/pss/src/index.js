export {
  createStyles as default
} from './core'

export * from './constants'
export * from './core'
export * from './mixins'
export * from './styles'
export * from './utils'
export * from './values'
