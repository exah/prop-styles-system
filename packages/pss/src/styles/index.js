export * from './atomic'
export * from './border-color'
export * from './border-radius'
export * from './border'
export * from './box'
export * from './box-alignment'
export * from './box-shadow'
export * from './box-style'
export * from './colors'
export * from './display'
export * from './flex'
export * from './flex-container'
export * from './font-family'
export * from './font-size'
export * from './font-weight'
export * from './gap'
export * from './grid-container'
export * from './grid-item'
export * from './hide'
export * from './letter-spacing'
export * from './line-height'
export * from './opacity'
export * from './order'
export * from './outline'
export * from './overflow'
export * from './position'
export * from './position-offsets'
export * from './ratio'
export * from './sizes'
export * from './space'
export * from './text'
export * from './text-align'
export * from './text-color'
export * from './text-overflow'
export * from './text-style'
export * from './transform'
export * from './transition'
export * from './white-space'
export * from './z-index'
