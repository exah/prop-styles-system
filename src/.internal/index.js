export * from './direction-styles'
export * from './size-rule'
export * from './space-rule'
