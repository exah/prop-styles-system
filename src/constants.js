export const DEFAULT_KEY = 'default'
export const COLORS_KEY = 'color'
export const PALETTE_KEY = 'palette'
export const BORDER_KEY = 'border'
export const MEDIA_KEY = 'media'
export const SIZES_KEY = 'size'
export const SPACE_KEY = 'space'
export const RADIUS_KEY = 'radius'
export const TEXT_STYLE_KEY = 'textStyle'
export const FONT_KEY = 'fontFamily'
export const ALL_MEDIA_KEY = 'all'

export const CSS_DEFAULT_VALUE = 'unset'

export const CSS_PROPS_DEFAULTS = {
  color: 'inherit',
  borderColor: 'transparent',
  backgroundColor: 'transparent'
}

export const DEFAULT_THEME_MEDIA = {
  [DEFAULT_KEY]: null
}

export const DEFAULT_THEME_SPACE = {
  [DEFAULT_KEY]: [ 0 ]
}
