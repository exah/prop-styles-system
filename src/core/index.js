export * from './create-prop-styles'
export * from './every-media'
export * from './media-rule'
export * from './rule'
