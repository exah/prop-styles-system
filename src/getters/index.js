import { fallbackTo, path, identity, mapObj } from '@exah/utils'
import { hasMediaKeys, keys } from '../utils'

import {
  DEFAULT_KEY,
  MEDIA_KEY,
  DEFAULT_THEME_MEDIA
} from '../constants'

export const getTheme = (props) => (props && props.theme) || Object(props)

/**
 * ```js
 * import { themePath } from 'pss'
 * ```
 *
 * Get value from theme.
 *
 * @example
 * import { themePath } from 'pss'
 *
 * const Box = styled.div`
 *   width: ${themePath('size.card')};
 *   background-color: ${themePath('color.red', 'hotpink')};
 * `
 *
 * <Box /> // → { width: 200px; background-color: hotpink; }
 */

export const themePath = (input, defaultValue) => (props) => path(
  input,
  defaultValue
)(getTheme(props))

/**
 * ```js
 * import { mq } from 'pss'
 * ```
 *
 * Get value from theme.
 *
 * @param [mediaKey] — key in `theme.media`
 *
 * @example
 * import { mq, themePath } from 'pss'
 *
 * const Box = styled.div`
 *   ${mq('sm')} {
 *     background-color: ${themePath('color.red', 'hotpink')};
 *   }
 * `
 *
 * <Box /> // → @media (max-width: 600px) { background-color: red; }
 */

export const mq = (mediaKey = 'default') =>
  (props) => `@media ${getMedia(mediaKey)(props)}`

export const getThemeMedia = (props) => ({
  ...DEFAULT_THEME_MEDIA,
  ...path(MEDIA_KEY)(getTheme(props))
})

export const getDefault = (input, defaultValue = DEFAULT_KEY) => themePath(
  [ DEFAULT_KEY, input ],
  defaultValue
)

export const getMedia = (input, media) => media
  ? path(input)(media)
  : (props) => path(input)(getThemeMedia(props))

export const getThemeValue = (
  themeDataKey,
  transformValue = identity
) => (
  input,
  defaultValue,
  defaultMediaKey
) => (props) => {
  const themeKey = input === true
    ? getDefault(themeDataKey)(props)
    : input

  const themeData = themePath(themeDataKey)(props)
  const themeValue = path(themeKey)(themeData)

  if (Object(themeValue).hasOwnProperty(defaultMediaKey)) {
    return transformValue(themeValue[defaultMediaKey])
  }

  if (hasMediaKeys(getThemeMedia(props), keys(themeValue))) {
    return mapObj(
      (key, value) => ({ [key]: transformValue(themeValue[key]) }),
      themeValue
    )
  }

  return transformValue(fallbackTo(themeValue, defaultValue))
}
