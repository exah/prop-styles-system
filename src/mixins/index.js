export * from './mq'
export * from './prop'
