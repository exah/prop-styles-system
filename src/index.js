// @flow

import * as styles from './styles'

export {
  createColor,
  createPaletteStyle,
  createPropStyles as default,
  createPropStyles as pss,
  createPropStyles,
  createRule,
  createSize,
  createSizeMixin,
  createSpace,
  createSpaceMixin,
  createThemeStyles,
  everyMedia
} from './core'

export {
  getColor,
  getColors,
  getPalette,
  getSize,
  getSpace,
  themePath
} from './getters'

export {
  base,
  border,
  colors,
  cssProp,
  display,
  ellipsis,
  flex,
  flexItem,
  float,
  fontFamily,
  marginPropStyles,
  overflow,
  paddingPropStyles,
  position,
  ratio,
  sizes,
  space,
  system,
  text,
  textStyle,
  typography,
  utility
} from './prop-styles'

export {
  combineSelectors as cs,
  combineSelectors,
  mediaPropSelector as mps,
  mediaPropSelector,
  propSelector as ps,
  propSelector,
  themeSelector as ts,
  themeSelector
} from './selectors'

export {
  styles
}

export {
  combine,
  sizeValue,
  skipPropValue,
  spaceValue,
  wrap,
  wrapIfMedia
} from './utils'
