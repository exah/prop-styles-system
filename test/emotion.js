import test from 'ava'
import renderer from 'react-test-renderer'
import { createElement as h } from 'react'
import { css, caches, flush } from 'emotion'
import styled from 'react-emotion'
import { space, sizes, colors } from '../src'

const theme = {
  media: {
    M: '(max-width: 600px)'
  },
  space: [ 0, 8, 16, 32, 64 ],
  palette: {
    default: {
      bg: '#ffffff',
      fg: '#000000'
    },
    inverted: {
      bg: '#000000',
      fg: '#ffffff'
    }
  }
}

test('basic', (t) => {
  flush()

  const sizeClassName = css(sizes({ theme, width: true, maxHeight: (3 / 4) }))

  t.is(Object.keys(caches.registered).includes(sizeClassName), true)
  t.snapshot(caches.registered)

  flush()

  const spaceClassName = css(space({ theme, mg: true }))

  t.is(Object.keys(caches.registered).includes(spaceClassName), true)
  t.snapshot(caches.registered)

  flush()

  const tmClassName = css(colors({ theme, tm: true }))

  t.is(Object.keys(caches.registered).includes(tmClassName), true)
  t.snapshot(caches.registered)

  flush()

  const Box = styled('div')(space, sizes, colors)
  const element = h(Box, { theme, width: true, tm: true, mg: { all: true, M: 0 } })
  const tree = renderer.create(element).toJSON()

  t.snapshot(tree)
  t.snapshot(caches.registered)
})
