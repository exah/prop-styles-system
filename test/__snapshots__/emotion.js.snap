// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
.emotion-0 {
  margin: 8px;
  width: 100%;
  background-color: #ffffff;
  color: #000000;
}

@media (max-width:600px) {
  .emotion-0 {
    margin: 0;
  }
}

<div
  className="emotion-0"
  width={true}
/>
`;
